UsageV3 S360 Updates

JoeBeach-msft opened the pull request:
## ARM API Information (Control Plane)

<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow.
</i>
Azure 1st Party Service can try out the [Shift Left](https://aka.ms/ShiftLeft) experience to initiate API design review from ADO code repo.  If you are interested, may request engineering support by filling in with the form https://aka.ms/ShiftLeftSupportForm.

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] update existing version for new feature
      - [ ] new service onboarding
      - [ ] new API version
      - [x] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. By default, Azure SDKs of all languages (.NET/Python/Java/JavaScript for both management-plane SDK and data-plane SDK, Go for management-plane SDK only ) MUST be refreshed with/after swagger of new version is published. If you prefer NOT to refresh any specific SDK language upon swagger updates in the current PR, please leave details with justification here.

### Contribution checklist (MS Employees Only):
- [ ] I commit to follow the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy) of "no breaking changes"
- [ ] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [ ] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist

> **Applicability**: :warning:
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs
> - Adding new properties
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [ ] Check this box if any of the following apply to the PR so that the label "ARMReview" and "WaitForARMFeedback" will be added by bot to kick off ARM API Review.  Missing to check this box in the following scenario may result in delays to the ARM manifest review and deployment.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits. You can use OpenAPIHub to initialize the PR for adding a new version. For more details refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [ ] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them.

### Breaking Change Review Checklist
If you have any breaking changes as defined in the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy/), request approval from the Breaking Change Review Board.

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Additional details on the process and office hours are on the [Breaking Change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

NOTE: To update API(s) in public preview for over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).


# UsageV3 S360 Updates

Added definitions for 202 responses and added 'error' as a property to definition for 'ErrorResponse'.

These changes were made to correct the below three items that were flagging for swagger compliance in S360.

![image](https://user-images.githubusercontent.com/125317918/219100738-d83574b8-e904-4d26-a66a-3ba2d6d8b89e.png)
![image](https://user-images.githubusercontent.com/125317918/219100889-24aba820-7ab1-4545-a6d4-1ce041c81dfa.png)
![image](https://user-images.githubusercontent.com/125317918/219100948-4cb2c51b-e6ce-44d1-b1e1-fa723641a11f.png)

JoeBeach-msft committed:
UsageV3 S360 Updates

Corrected three items that were flagging for swagger compliance in S360.

Added definitions for 202 responses and added 'error' as a property to definition for 'ErrorResponse'.

peteokos commented on specification/commerce/resource-manager/Microsoft.Commerce/preview/2015-06-01-preview/commerce.json:
```diff
@@ -137,6 +143,12 @@
             }
           },
+          "202": {
+            "description": "Response indicating that a request has been accepted for processing. However, the processing has not been completed.",
```
You should not make any change to the ratecard block of the swagger, ratecard is a different service and I don't think they support 202 response code

JoeBeach-msft committed:
Removed 202 Response for Ratecard

Removed the 202 response for Ratecard

Ratecard is a different service that is not getting flagged for in S360 and does not support 202 response code.

JoeBeach-msft committed:
Undo ErrorMessage Properties

Swagger will not let ErrorResponse contain an error property with subproperties.

Reverting ErrorResponse to contain "code" and "message" as its properties.

JoeBeach-msft committed:
Ran Prettier

Ran prettier in order to address Pull Request check.

JoeBeach-msft committed:
Added xm-ms-long-running-operation Property

Added xm-ms-long-running-operation Property to the UsageAggregates_List get call in order to justify the addition of the 202 response.

JoeBeach-msft committed:
Added x-ms-long-running-operation-options

Added x-ms-long-running-operation-options.

Adding a 202 response requires that these options be set.

JoeBeach-msft committed:
Prettier Run

Reran prettier after the last commit to correct formatting issues.

JoeBeach-msft committed:
Added x-ms-example for GetUsageAggregates_List

Added x-ms-example for GetUsageAggregates_List. PR check was flagging because there was no example present. Will create the referenced json file in the next commit.

JoeBeach-msft committed:
Added example file for UsageAggregates List

There was previously no example for UsageAggregates List, but there was one for RateCard.

The following documentation was used to generate this example file.
https://learn.microsoft.com/en-us/previous-versions/azure/reference/mt219001(v=azure.100)?redirectedfrom=MSDN#sample-responses

JoeBeach-msft committed:
Corrected Formatting for UsageAggregatesList Example

Value Array was missing brackets for its element.

JoeBeach-msft committed:
Corrected Example File Name

Corrected the file name to exclude an underscore.

JoeBeach-msft committed:
Added Location Header to 202 Response

Added location header to the 202 response.

JoeBeach-msft committed:
Corrected Location to use double quotes

Corrected formatting issue with the previous commit that added a location header to the 202 response.

JoeBeach-msft committed:
Updated Example Fields

Updated infofields to match the example given in the api documentation.

JoeBeach-msft committed:
Added 202 Response

Added a 202 response to the example file.

JoeBeach-msft committed:
GetUsageAggregatesList.json Prettier Update

Reran Prettier script on GetUsageAggregatesList.json

JoeBeach-msft committed:
Updated case 'Daily' and Example 202 Response

Updated case 'Daily' and Example 202 Response. These changes were made previously, but lost at some point.

JoeBeach-msft committed:
Removed Error from 202 Body

Error property not allowed in the response.

NetRock commented on specification/commerce/resource-manager/Microsoft.Commerce/preview/2015-06-01-preview/commerce.json:
```diff
@@ -151,6 +172,14 @@
   "definitions": {
     "InfoField": {
       "properties": {
+        "meteredRegion": {
```
these are new properties.
Need to be a new api-version

JoeBeach-msft replied:
These are actually existing properties that were incorrectly missing in the Swagger document. We did not update the api, so a new api version is not required.

These properties can be seen in the first sample response for the current api version:
https://learn.microsoft.com/en-us/previous-versions/azure/reference/mt219001(v=azure.100)?redirectedfrom=MSDN#response-codes

JoeBeach-msft replied:
Just to be explicitly clear. Our team did not in any way update the way that the Api works. The swagger document is being updated to match its current behavior.

JoeBeach-msft replied:
Upon further review, it looks like our API returns a blank response for Infofields, so these properties are no longer used. I will remove them from the example and Commerce.json.

JoeBeach-msft committed:
Removed InfoFields properties.

Removed all subproperties for InfoFields. Our api no longer returns any properties for InfoFields.

JoeBeach-msft committed:
Removed Infofields properties.

Removed Infofields properties. Our api no longer returns any of these properties in a 200 response.

JoeBeach-msft committed:
Removed InfoFields Properties

Removed Property from InfoFields. This should have been removed along with the other properties in the previous commit.

xiaoxuqi-ms commented on specification/commerce/resource-manager/Microsoft.Commerce/preview/2015-06-01-preview/commerce.json:
```diff
@@ -27,6 +27,16 @@
+        "x-ms-long-running-operation-options": {
+          "final-state-via": "location",
+          "final-state-schema": "#/definitions/UsageAggregationListResult"
+        },
```
Actually, this is not a long running operation, and it should not be marked as LRO, so please remove this .

pilor replied:
yes, please do remove x-ms-long-running-operation

JoeBeach-msft replied:
I removed the long running operation attribute.

Just for background: I originally added this because the linter told me that this was required for the 202. Removing this might generate another linter flag.

xiaoxuqi-ms commented on specification/commerce/resource-manager/Microsoft.Commerce/preview/2015-06-01-preview/commerce.json:
```diff
@@ -88,6 +98,17 @@
             }
           },
+          "202": {
+            "description": "Response indicating that a request has been accepted for processing. However, the processing has not been completed.",
```
Hi @pilor , thanks for your review, Joe raised this pr to fixed an S360 item which complains 202 is not defined. As we discussed offline, although 202 is not allowed in GET apis, but this behavior was released 2014. https://learn.microsoft.com/en-us/previous-versions/azure/reference/mt219001(v=azure.100)?redirectedfrom=MSDN. Could you kindly help to approve this change?

pilor replied:
please remove the long running operation attribute

JoeBeach-msft replied:
I removed the long running operation attribute.

Just for background: I originally added this because the linter told me that this was required for the 202. I think removing this will generate another alert.

xiaoxuqi-ms replied:
@weidongxu-microsoft , this api is not a LRO but the service takes several hours to aggregate the data to returned, 202 is returned since the api was released in year2014

weidongxu-microsoft replied:
Even we allow 202, why it is ErrorResponse?

HTTP client will timeout in a few minutes. If API takes hours but not an LRO, it just means it won't return anything useful.

Fine with it. Java likely will just refuse release anything unless this solved.

JoeBeach-msft committed:
Removed LongRunning operation

Removed longrunning operation from UsageAggregates_List.

pilor commented on specification/commerce/resource-manager/Microsoft.Commerce/preview/2015-06-01-preview/commerce.json:
```diff
@@ -88,6 +93,17 @@
+          "202": {
+            "description": "Response indicating that a request has been accepted for processing. However, the processing has not been completed.",
+            "headers": {
+              "Location": {
```
>Location [](http://example.com/codeflow?start=15&length=8)

Do you actually return a location header?  If not, remove this

JoeBeach-msft replied:
Removed.

JoeBeach-msft committed:
Removed Location Header

Location header no longer needed.

weidongxu-microsoft commented on specification/commerce/resource-manager/Microsoft.Commerce/preview/2015-06-01-preview/commerce.json:
```diff
@@ -88,6 +93,12 @@
+          "202": {
+            "description": "Response indicating that a request has been accepted for processing. However, the processing has not been completed.",
+            "schema": {
+              "$ref": "#/definitions/ErrorResponse"
```
This 202 is wrong.

Why 202 returns ErrorResponse?

weidongxu-microsoft replied:
And what 202 do for a GET?
